src/dependency_injector/providers.pyx

```diff
@@ -2928,6 +2934,89 @@ cdef class ThreadLocalSingleton(BaseSingleton):
             future_result.set_result(instance)
 
 
+cdef class ContextLocalSingleton(BaseSingleton):
+    """Context-local singleton provides single objects in scope of a context.
+
+    .. py:attribute:: provided_type
+
+        If provided type is defined, provider checks that providing class is
+        its subclass.
+
+        :type: type | None
+
+    .. py:attribute:: cls
+       :noindex:
+
+        Class that provides object.
+        Alias for :py:attr:`provides`.
+
+        :type: type
+    """
+    _none = object()
+
+    def __init__(self, provides=None, *args, **kwargs):
+        """Initializer.
+
+        :param provides: Provided type.
+        :type provides: type
+        """
+        if not contextvars:
+            raise RuntimeError(
+                'Contextvars library not found. This provider '
+                'requires Python 3.7 or a backport of contextvars.'
+            )
+
+        super(ContextLocalSingleton, self).__init__(provides, *args, **kwargs)
+        self.__storage = contextvars.ContextVar('__storage', default=self._none)
+
+    def reset(self):
+        """Reset cached instance, if any.
+
+        :rtype: None
+        """
+
```

Comment:
Could you please remove this empty line? I usually try to avoid them.

Reply:
Sure